goto.py

```diff
@@ -164,37 +177,38 @@ def _patch_code(code):
         target_depth = len(target_stack)
         if origin_stack[:target_depth] != target_stack:
             raise SyntaxError('Jump into different block')
+        
+        size = 0
+        for i in range(len(origin_stack) - target_depth):
+            size += _get_instruction_size('POP_BLOCK')
+        size += _get_instruction_size('JUMP_ABSOLUTE', target // _BYTECODE.jump_unit)
+        
+        moved_to_end = False
+        if pos + size > end:
+            # not enough space, add at end
+            pos = _write_instruction(buf, pos, 'JUMP_ABSOLUTE', len(buf) // _BYTECODE.jump_unit)
+            
+            if pos > end:
+                raise SyntaxError('Goto in an incredibly huge function') # not sure if reachable
+            
+            size += _get_instruction_size('JUMP_ABSOLUTE', end // _BYTECODE.jump_unit)
+            
+            moved_to_end = True
+            pos = len(buf)
+            buf.extend([0] * size)
 
-        failed = False
         try:
             for i in range(len(origin_stack) - target_depth):
                 pos = _write_instruction(buf, pos, 'POP_BLOCK')
-
-            if target >= end:
-                rel_target = (target - pos) // _BYTECODE.jump_unit
-                oparg_bits = 0
-
-                while True:
-                    rel_target -= (1 + _BYTECODE.argument.size) // _BYTECODE.jump_unit
-                    if rel_target >> oparg_bits == 0:
-                        pos = _write_instruction(buf, pos, 'EXTENDED_ARG', 0)
-                        break
-
-                    oparg_bits += _BYTECODE.argument_bits
-                    if rel_target >> oparg_bits == 0:
-                        break
-
-                pos = _write_instruction(buf, pos, 'JUMP_FORWARD', rel_target)
-            else:
-                pos = _write_instruction(buf, pos, 'JUMP_ABSOLUTE', target // _BYTECODE.jump_unit)
-
-        except (IndexError, struct.error):
-            failed = True
-
-        if failed or pos > end:
-            raise SyntaxError('Jump out of too many nested blocks')
-
-        _inject_nop_sled(buf, pos, end)
+            pos = _write_instruction(buf, pos, 'JUMP_ABSOLUTE', target // _BYTECODE.jump_unit)
+        except (IndexError, struct.error) as e:
+            raise SyntaxError("Internal error", e)
```

Comment:
I get an exception when trying to pass another exception as second argument to the `SyntaxError` constructor:

```
>>> SyntaxError('foo', Exception())
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'Exception' object is not iterable
```

Reply:
Oh right. Serves me right for adding code that shouldn't be reachable. (It's since removed in the other pull requests, I believe)

Reply:
This code was removed in the refactor I added to this PR, since it shouldn't be reachable - we now always check if there is space in the buffer before writing to it.